stream_alert/rule_processor/rules_engine.py

```diff
@@ -380,20 +380,24 @@ def process(cls, input_payload):
                 matcher_result = cls.match_event(record, rule)
                 if not matcher_result:
                     continue
+
+                types_result = None
                 if rule.datatypes:
                     types_result = cls.match_types(record,
                                                    payload.normalized_types,
                                                    rule.datatypes)
-                    record['normalized_types'] = types_result
 
+                record_copy = record.copy()
```

Comment:
I don't love that we're copying a record in memory for every rule in the loop. Can't we just see if the the `record` has the `normalized_types` key in it and delete it if it does? For example, around line 374 above, can you just:
```
if 'normalized_types' in record:
    del record['normalized_types']
```

Reply:
We can not use `del` method here because once call to delete `normalizied_types` from a record, it will also delete `normalized_types` from alert['record'] who has `normalized_types` field. 
For a trade-off, I will make a record copy only when normalized types take effect.